File: base64.py
alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
resultArray = bytearray()
byteArray = bytearray()
byteString = ''

def base64_encode(inputString):

    byteArray = bytearray(inputString, 'utf-8')
    sumOfBit = len(byteArray)*8   #Here we are finding sum of bits that % 6 == 0, until that we add null bytes
    while(sumOfBit%6 != 0):
        byteArray.append(0)
        sumOfBit+=8

    byteString = ''
    for byte in byteArray:  #append bytes to 8 bits and then convert bytes => string of bytes
        number = "{0:b}".format(byte)
        byteString+= '0'*(8-len(number)) + number

    while(byteString):  #we take every 6 bits and convert them to bytes
        resultArray.append(int(byteString[0:6],2))
        byteString=byteString[6:]

    result=''
    for byte in resultArray:    #then we are building string from our alphabet
        if byte==0:
            result += '='
        else:
            result += alphabet[byte]

    return result

def base64_decode(inputString):
    for letter in inputString:
        if letter=='=':
            byte=0
        else:
            byte = alphabet.index(letter)
        byteArray.append(byte)

    byteString=''
    for byte in byteArray:
        number = "{0:b}".format(byte)
        byteString+='0'*(6-len(number)) + number

    while(byteString):
        resultArray.append(int(byteString[0:8],2))
        byteString=byteString[8:]

    return resultArray.decode("utf-8")

def main():
    print("Welcome to base64 encoder/decoder. Input your string:",end=' ')
    string = input()
    print("1) encode\n2) decode\nWhat do you want:",end=' ')
    ans = input()
    if ans=='1':
        print("Encoded string: " + base64_encode(string))
    elif ans=='2':
        print("Decoded string: " + base64_decode(string))
    else:
        print("Bad answer, exiting.")

if __name__ == "__main__":
    main()

File: base64.cpp
#include <iostream>
#include <string>
using namespace std;

const string base64_chars =
"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
"abcdefghijklmnopqrstuvwxyz"
"0123456789+/";

string base64encode(string text) {
	string result;
	int a,b,c;
	int len = text.length();
	int nullbytes = 3-len%3;
	if(nullbytes==3) 
		nullbytes=0;
	
	for(int i = 0; i<nullbytes; i++)
		text+='\0';

	for (int i = 0; i < len; i+=3) {
		a=text[i];
		b=text[i+1];
		c=text[i+2];
		result += base64_chars[a >> 2];
		result += base64_chars[(((a & 0x3) << 6) + (b >> 2)) >> 2];
		result += base64_chars[((b & 0xF) << 2) + (c >> 6)];
		result += base64_chars[c & 0x3F];		
	}

	result.resize(result.length()-nullbytes);
	for (int i = 0; i < nullbytes; i++)
		result += '=';

	return result;
}
string base64decode(string text) {
	string result;
	int a, b, c, d;
	int len = text.length();
	int nullbytes = 0;

	for (int i = 0; i<len; i++) {
		if (text[i] == '=')
			nullbytes++;
		text[i]=char(base64_chars.find(text[i]));			
	}
		
	for (int i = 0; i < len; i += 4) {
		a = text[i];
		b = text[i + 1];
		c = text[i + 2];
		d = text[i + 3];
		result += (a << 2) + (b >> 4);
		result += ((b&0xF)<<4) + (c>>2);
		result += ((c&0x3)<<6)+d;		
	}

	result.resize(result.length() - nullbytes);
	return result;
}

int main(int argc, char* argv[]) {
	
	string msg = "Base64 encoder/decoder\nHow to use:\n\t-e <text>\t-\tencode text\n\t-d <text>\t-\tdecode text";

	if (argc != 3) {
		cout<<msg;
		return -1;
	}	
	string flag = argv[1];
	string text = argv[2];

    if(flag =="-e")
        cout<<"base64: "<<base64encode(text);
	else if(flag =="-d")
		cout<<"decoded text:"<<base64decode(text);
	else {
		cout<<msg;
		return -2;	
	}
	return 0;
}